File: src/leetCodeProblems/Greedy/MinAddToMakeValidParentheses921.java
package leetCodeProblems.GreedyAlgo;

/**
 * LeetCode - https://leetcode.com/problems/minimum-add-to-make-parentheses-valid/
 */

public class MinAddToMakeValidParentheses921 {

    public int minAddToMakeValid(String s) {
        int c_need = 0;
        int ans = 0;

        for (int i = 0; i < s.length(); i++) {

            if (s.charAt(i) == '(') {
                c_need += 1;
            }
            else {
                c_need -= 1;

                // This is TRICKY and IMPORTANT CODE. To reset the variables. This is the only closing bracket.
                if (c_need == -1) {
                    c_need = 0;
                    ans++; // an open bracket is needed here. Hence increment "ans".
                }
            }
        }

        return c_need + ans;
    }

    public static void main(String[] args) {

        MinAddToMakeValidParentheses921 obj = new MinAddToMakeValidParentheses921();

        System.out.println(obj.minAddToMakeValid("())"));
        System.out.println(obj.minAddToMakeValid("((("));
    }
}

File: src/leetCodeProblems/StacksAndQueues/MinAddToMakeValidParentheses921.go
package main

/*
- Leetcode - https://leetcode.com/problems/minimum-add-to-make-parentheses-valid/description/
- Space - O(n)
- Time - O(n)
*/

type stackParenthesis []rune

func (s *stackParenthesis) pop() rune {
	temp := (*s)[len(*s)-1]
	*s = (*s)[:len(*s)-1]
	return temp
}

func (s *stackParenthesis) push(r rune) {
	*s = append(*s, r)
}

func (s stackParenthesis) isEmpty() bool {
	return len(s) == 0
}

func (s stackParenthesis) peek() rune {
	return s[len(s)-1]
}

func (s stackParenthesis) len() int {
	return len(s)
}

func minAddToMakeValid(s string) int {
	st := new(stackParenthesis)

	for _, r := range s {
		if st.isEmpty() {
			st.push(r)
		} else {

			if string(r) == ")" && string(st.peek()) == "(" {
				st.pop()
			} else {
				st.push(r)
			}
		}
	}

	return st.len()
}

// func main() {
// 	log.Println(minAddToMakeValid("())"))
// }
